### ISSUE TITLE
Moving to Alectryon : first draft

### ISSUE THREAD

**Issue by Casteran** [coq-community/hydra-battles#70]
@start974 @Zimmi48 @palmskog @cpitclaudel 

The part of hydra-battles's documentation dedicated to ordinals is now made with Alectryon.
(sniipets extracted from theories/ordinals/*/*.v )

This document is stiil a draft, and I will re-read it slowly now. After the corrections, it would be nice to make a release.

I noticed two issues:
  - sometimes, the implication arrow `->` is broken  into two pieces (in long formulas) . 
  - There were also big vertical spaces between two adjacent snippets.

Many thanks to all of you !

**Comment by cpitclaudel** [comment]
I looked a bit further, and it does sound like we would need some syntax for changing the default rendering past for part of a snippet, since there are many uses of that feature.

I will think about it.

**Comment by cpitclaudel** [comment]
OK, implementation is in #71. 

> I will think about it.

I went the easiest way possible: I implemented `(* begin snippet endOfProof:: no-out *)` headers, and I further subdivide snippets around comments of the form `(*| .. coq:: … |*)`, keeping track of the relevant output customizations for each sub-snippet; they are all reassembled in the end..

**Comment by Casteran** [comment]
I noted a small issue: I wanted to implement the pattern "rest of proof skipped", for instance through the snippet 
mGeb (in file theories/ordinals/Hydra/Omega_Small.v) I couldn't unstick the final `Qed`from the comment `(* ... *)` in the pdf (p. 45). I tried a workaround: to add a snippet just for the final `Qed`.   More generally, A possible way to show only parts of a proof script would  be to use a sequence of snippets instead of a single one with sub-snippets in `none`mode ?

The issue is visible on the last commited version.

**Comment by cpitclaudel** [comment]
Thanks, that's a mistake I made in the coalesce_text function in Alectryon (I wrote it a while ago but didn't use it until this project).  And indeed after the fix:

```diff
diff --git a/snippets.broken-concat/Omega2_Small/mGeb.tex b/snippets/Omega2_Small/mGeb.tex
index 0e829ab..00e1e78 100644
--- a/snippets.broken-concat/Omega2_Small/mGeb.tex
+++ b/snippets/Omega2_Small/mGeb.tex
@@ -2,12 +2,13 @@
   % Generator: Alectryon
   \sep
   \begin{txt}
-    ~~~~\PY{c}{(*~...~*)}
+    ~~~~\PY{c}{(*~...~*)}\nl
+    ~~~~~~\nl
   \end{txt}
   \sep
   \begin{sentence}
     \begin{input}
-      \PY{k+kn}{Qed}\PY{o}{.}
+      ~~\PY{k+kn}{Qed}\PY{o}{.}
     \end{input}
   \end{sentence}
 \end{alectryon}
```

**Comment by cpitclaudel** [comment]
> More generally, A possible way to show only parts of a proof script would be to use a sequence of snippets instead of a single one with sub-snippets in nonemode ?

Yes, but there will be problems with spacing, since you'll have two snippets and hence two `alectryon` environments.

I do think it would be cleaner to have multiple snippets with no `.. coq:: none` comments in the middle of them, tough (it would further simplify the driver, too).  If you want to go that route, then we would change the LaTeX slightly: you would have a command `\includesnippets{snippet1,snippet2,snippet3}` instead of a plain `\include`, and that command would include all three snippets in a single `\alectryon` block.

Let me know if you want to go that route.


 then we would change the LaTeX

**Comment by Casteran** [comment]
I like the idea of including multiple snippets ! Besides the issue of spacing, I believe it will make the latex source more readable.

**Comment by cpitclaudel** [comment]
> sometimes, the implication arrow -> is broken into two pieces (in long formulas) .

Fixed in #72

**Comment by cpitclaudel** [comment]
> I like the idea of including multiple snippets !

OK, prototype in #73

**Comment by Casteran** [comment]
@Zimmi48 The last version I pushed fails almost immediately on NIX CI on the "Checking presence of CI target coq" step.
Is this just a problem of the machine which runs CI ? Or a mistake of mine ?

**Comment by Zimmi48** [comment]
The error is surprising, especially because it doesn't give any useful output to debug it. It's likely though that this will be a transient error so let's wait to see if it persists.

**Comment by Casteran** [comment]
> > Perhaps we should also provide a tar file of the html files, with instructions for making the links `../theories/html` correct ?
> 
> Indeed, we can provide a separate archive with HTML files. But in this case, I suggest manually preprocessing the HTML files to fix any linking issues before attaching this archive. Even better would in my opinion be that the main source archive contains the manually preprocessed HTML files (correct links) along with Coq source files.

So, What you suggest is to generate (by `coqdoc -g -html Foo.v`) a file `Foo.html`which will be in the same directory as `Foo.v`?

**Comment by palmskog** [comment]
> So, What you suggest is to generate (by `coqdoc -g -html Foo.v`) a file `Foo.html`which will be in the same directory as `Foo.v`?

Recall that a GitHub tag/release generates a tarball with the contents of the repo (possibly excepting some files). This means the user has to run `make html` manually to get the HTML files in `theories/html`. I was suggesting to prepare a release archive/tarball manually (in addition to the generated one) to have these HTML files already included in the usual directory.

**Comment by Casteran** [comment]
Coqdoc generation fails on NIX CI with a message I don't understand. (works on my machine)

File "./theories/ordinals/OrdinalNotations/ON_Omega_plus_omega.v", line 6, characters 0-78:
Error:
/home/runner/work/hydra-battles/hydra-battles/theories/ordinals/OrdinalNotations/ON_Finite.vo: premature end of file. Try to rebuild it.

"coqc"  -q -w -notation-overridden -w -ambiguous-paths -w -deprecated-instance-without-locality -w -deprecated-hint-rewrite-without-locality  -R theories/ordinals hydras -R theories/additions additions -R theories/gaia gaia_hydras -R exercises hydra_exercises theories/ordinals/OrdinalNotations/ON_plus.v
make[1]: *** [Makefile.coq:724: theories/ordinals/OrdinalNotations/ON_Omega_plus_omega.glob] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/home/runner/work/hydra-battles/hydra-battles'
make: *** [Makefile:5: html] Error 2
Error: Process completed with exit code 2.

**Comment by Casteran** [comment]
It works now!  All the chapters of the book use Alectryon. 
There are still a few verbatim (source and answers) when the computations are too long or result in errors during the snippet extraction.
I have now to look for typos, overfull boxes, ... ¡ Mañana !

**Comment by cpitclaudel** [comment]
!! :tada: :partying_face:
